src/http.js

```diff
@@ -36,56 +56,43 @@ class Http {
       return Promise.resolve();
     }
 
-    let getToken;
-    if (this.accessToken) {
-      getToken = Promise.resolve(this.accessToken);
-    } else {
-      getToken = this.oauth.getAccessToken()
-        .then(this.errorFilter)
-        .then(res => res.json())
-        .then(accessToken => {
-          this.accessToken = accessToken;
-          return accessToken;
-        });
-    }
-
-    return getToken
+    return this.oauth.getAccessToken()
         .then(accessToken => {
           request.headers.Authorization = `Bearer ${accessToken.access_token}`;
         });
   }
 
-  errorFilter(response) {
-    if (response.status >= 200 && response.status < 300) {
-      return response;
-    } else {
-      return response.text()
-      .then(body => {
-        let err;
-
-        // If the response is JSON, assume it's an Okta API error. Otherwise, assume it's some other HTTP error
-
-        try {
-          err = new OktaApiError(response.url, response.status, JSON.parse(body), response.headers);
-        } catch (e) {
-          err = new HttpError(response.url, response.status, body, response.headers);
-        }
-        throw err;
-      });
-    }
-  }
-
   http(uri, request, context) {
     request = request || {};
     context = context || {};
     request.url = uri;
     request.headers = Object.assign(this.defaultHeaders, request.headers);
     request.method = request.method || 'get';
+
+    const promise = this.prepareRequest(request)
+      .then(() => this.requestExecutor.fetch(request))
+      .then(Http.errorFilter)
+      .catch(error => {
+        // Handle oauth access token expiration
+        if (this.oauth && error && error.status === 401) {
+          return this.oauth.introspectAccessToken()
```

Comment:
> Do we need to introspect the token? Would it be simpler to just assume it's invalid, clear it from local cache, and then try to get a new access token?

Since there are other scenarios can also cause `status === 401`, I added introspect here just want to make sure it retry on getting access token only when token is not active. As it's not a call will happen frequently, and it also eliminates other retry cases. But I agree, **clearCachedToken should happen as long as error happen.**

> Does this logic work OK if I submit bad JWKs data? In other words, if getting an access token fails because of bad credentials, does it give a 400 or 401 response? If it's 401 it seems like we could get in a failure loop?

ErrorFilter had been added for getAccessToken call (https://github.com/okta/okta-sdk-nodejs/pull/138/files#diff-3a1276f4dc4422026a36b4b8750287bfR52) to throw the error here. Actually `introspectAccessToken ` can also help in this case to identity current token status.

Reply:
Moved `clearCachedAccessToken` call outside of `introspectAccessToken` to make sure all 401 error will clear cached token.

Reply:
Updated to only retry one time.